<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Modesto">
<meta name="dcterms.date" content="2024-04-26">

<title>Thermus TnSeq - Thermus Pangenome</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>


</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Thermus TnSeq</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-contents" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Contents</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-contents">    
        <li>
    <a class="dropdown-item" href="./reads_processing.html">
 <span class="dropdown-text">TnSeq Illumina Data Processing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./genes_classification.html">
 <span class="dropdown-text">TnSeq Genes classification</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./panthermus.html">
 <span class="dropdown-text">Thermus Pangenome</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
    <a href="https://github.com/mredrejo/thermus_tnseq" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#genomes-dataset-and-pangenome-construction" id="toc-genomes-dataset-and-pangenome-construction" class="nav-link active" data-scroll-target="#genomes-dataset-and-pangenome-construction"><span class="header-section-number">1</span> Genomes dataset and pangenome construction</a></li>
  <li><a href="#pangenomes-stats" id="toc-pangenomes-stats" class="nav-link" data-scroll-target="#pangenomes-stats"><span class="header-section-number">2</span> Pangenomes stats</a></li>
  <li><a href="#integration-of-tnseq-and-pangenomes" id="toc-integration-of-tnseq-and-pangenomes" class="nav-link" data-scroll-target="#integration-of-tnseq-and-pangenomes"><span class="header-section-number">3</span> Integration of TnSeq and pangenomes</a>
  <ul class="collapse">
  <li><a href="#correlation-between-tnseq-and-pangenome" id="toc-correlation-between-tnseq-and-pangenome" class="nav-link" data-scroll-target="#correlation-between-tnseq-and-pangenome">Correlation between TnSeq and Pangenome</a></li>
  <li><a href="#enrichment-analysis" id="toc-enrichment-analysis" class="nav-link" data-scroll-target="#enrichment-analysis">Enrichment analysis</a></li>
  </ul></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content column-page-right" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Thermus Pangenome</h1>
</div>


<div class="quarto-title-meta-author column-page-right">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Modesto </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Department of Biochemistry, UAM
          </p>
      </div>
  </div>

<div class="quarto-title-meta column-page-right">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">April 26, 2024</p>
    </div>
  </div>
  
    <div>
    <div class="quarto-title-meta-heading">Modified</div>
    <div class="quarto-title-meta-contents">
      <p class="date-modified">June 24, 2024</p>
    </div>
  </div>
    
  </div>
  


</header>


<section id="genomes-dataset-and-pangenome-construction" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Genomes dataset and pangenome construction</h1>
<p>We downloaded (May 27th, 2024) all the Phylum <em>Deinococcota</em> assemblies from Refseq database using the NCBI tool <code>datasets</code> (v.16.5.0) installed in Conda.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate ncbi_datasets</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> update <span class="at">-c</span> conda-forge ncbi-datasets-cli</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> thermaceae/thermus_refseq</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">datasets</span> download genome taxon Thermaceae <span class="at">--filename</span> thermaceae_dataset.zip <span class="at">--exclude-atypical</span> <span class="at">--assembly-source</span> <span class="st">'refseq'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>To obtain an index of the genome files, we converted the <code>dataset_catalog.json</code> to CSV (<a href="https://www.convertcsv.com/json-to-csv.htm" class="uri">https://www.convertcsv.com/json-to-csv.htm</a>) to obtain the file <a href="thermaceae/thermaceae_dataset/fasta.csv">fasta.csv</a>.</p>
<p>To work with homogeneous and updated genome annotations, we subsequently re-annotated all genome assemblies with <code>Bakta</code> (from Conda, v. 1.9.2) using the full database (DDBB v. 5.1) and the options <code>--skip-crispr --force</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> pangenomes/deinococcota/ncbi_dataset/data</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bakta</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> update bakta</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">bakta_db</span> download <span class="at">--output</span> /Volumes/Trastero4/ddbb <span class="at">--type</span> full </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co">#database is in external HD</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> <span class="va">IFS</span><span class="op">=</span>, <span class="bu">read</span> <span class="at">-r</span> <span class="va">col1</span> <span class="va">col2</span> <span class="va">col3</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">do</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>        <span class="ex">bakta</span> <span class="at">--db</span> /Volumes/Trastero4/ddbb/bakta-db <span class="at">--verbose</span> <span class="at">--output</span> ../../../bakta_results/<span class="va">$col3</span> <span class="at">--prefix</span> <span class="va">$col3</span> <span class="at">--locus-tag</span> <span class="va">$col3</span> <span class="at">--threads</span> 16 <span class="va">$col1</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">done</span> <span class="op">&lt;</span> <span class="op">&lt;(</span><span class="fu">tail</span> <span class="at">-n</span> +2 fasta.csv<span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The pangenomes were constructed with <code>PPanGGOLIN</code> (v. 2.0.5) installed in Conda. You can see the whole documentation about PPanGGOLIN and the output files <a href="https://ppanggolin.readthedocs.io/en/latest/index.html">here</a>. In order to construct different pangenomes at species, genus, family and order levels, we parse the taxonomy from NCBI datasets using <a href="https://www.ncbi.nlm.nih.gov/datasets/docs/v2/reference-docs/command-line/dataformat/"><code>dataformat</code></a> tool and then use a short R script (<a href="scripts/parse_taxonomy.R"><code>parse_taxonomy.R</code></a>) to generate the annotated genomes table lists.</p>
<p>After testing different alternative datasets and clustering combinations, we empirically set the MMSeqs clustering sequence identity and coverage parameters to 0.4 and 0.5, respectively.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#STEP 2</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Run ppanggolin and write extra output files</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bioconda</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> pangenomes</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="ex">ppanggolin</span> workflow <span class="at">--anno</span> deinococcota.gbff.list <span class="at">--basename</span> deinococcota <span class="at">--identity</span> 0.4 <span class="at">--coverage</span> 0.5 <span class="at">-o</span> deinococcota_i4c5 <span class="at">-c</span> 16 <span class="at">-f</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="ex">ppanggolin</span> workflow <span class="at">--anno</span> thermaceae.gbff.list <span class="at">--basename</span> thermaceae <span class="at">--identity</span> 0.4 <span class="at">--coverage</span> 0.5 <span class="at">-o</span> thermaceae_i4c5 <span class="at">-c</span> 16 <span class="at">-f</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="ex">ppanggolin</span> workflow <span class="at">--anno</span> thermus.gbff.list <span class="at">--basename</span> thermus <span class="at">--identity</span> 0.4 <span class="at">--coverage</span> 0.5 <span class="at">-o</span> thermus_i4c5 <span class="at">-c</span> 16 <span class="at">-f</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="ex">ppanggolin</span> workflow <span class="at">--anno</span> tt.gbff.list <span class="at">--basename</span> tthermophilus <span class="at">--identity</span> 0.4 <span class="at">--coverage</span> 0.5 <span class="at">-o</span> tt_i4c5 <span class="at">-c</span> 16 <span class="at">-f</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Additionally, before moving forward with the pangenome, as reference, we are going to incorporate the gene names from the <em>Thermus</em> <em>thermophilus</em> strain HB27, as annotated in the NCBI Refseq assembly (<a href="#0">GCF_000008125.1</a>).</p>
</section>
<section id="pangenomes-stats" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Pangenomes stats</h1>
<p>Now, we are going to have a look to the main pangenome stats.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>pangenomes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"tt_i4c5"</span>,<span class="st">"thermus_i4c5"</span>,<span class="st">"thermaceae_i4c5"</span>,<span class="st">"deinococcota_i4c5"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pangenomes) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Thermus thermophilus"</span>,<span class="st">"Thermus"</span>,<span class="st">"Thermaceae"</span>,<span class="st">"Deinococcota"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#stats from all metagenomes</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>cloud <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>shell <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>persistent <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pangenomes)){</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  cloud[i] <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">read_lines</span>(<span class="fu">paste0</span>(<span class="st">"pangenomes/"</span>,pangenomes[i],<span class="st">"/partitions/cloud.txt"</span>)))</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  shell[i] <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">read_lines</span>(<span class="fu">paste0</span>(<span class="st">"pangenomes/"</span>,pangenomes[i],<span class="st">"/partitions/shell.txt"</span>)))</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  persistent[i] <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">read_lines</span>(<span class="fu">paste0</span>(<span class="st">"pangenomes/"</span>,pangenomes[i],<span class="st">"/partitions/persistent.txt"</span>)))</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(cloud,shell,persistent)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(stats) <span class="ot">&lt;-</span> <span class="fu">names</span>(pangenomes)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">row.names</span>(stats),<span class="fu">stack</span>(stats))</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(stats) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"level"</span>,<span class="st">"numbers"</span>,<span class="st">"partition"</span>)</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(stats,<span class="fu">aes</span>(<span class="at">x=</span>level, <span class="at">y=</span>numbers,<span class="at">group=</span>partition)) <span class="sc">+</span> </span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>         <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">fill=</span>partition),<span class="at">stat =</span> <span class="st">"identity"</span>,<span class="at">position=</span><span class="st">"fill"</span>,<span class="at">color=</span><span class="st">"grey40"</span>, <span class="at">alpha=</span><span class="fl">0.8</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Gene families (%)"</span>)<span class="sc">+</span> <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>numbers),<span class="at">size=</span><span class="dv">5</span>, <span class="at">position =</span> <span class="fu">position_fill</span>(<span class="at">vjust=</span><span class="fl">0.5</span>) , <span class="at">col =</span> <span class="st">"black"</span>)<span class="sc">+</span><span class="fu">xlab</span>(<span class="st">"Pangenome level"</span>) <span class="sc">+</span><span class="fu">scale_y_continuous</span>(<span class="at">labels =</span> scales<span class="sc">::</span>percent) <span class="sc">+</span> <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="panthermus_files/figure-html/unnamed-chunk-1-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption><font size="2">Figure 20. Statistics of PPanGGOLIN pangenomes obtained with 40% identity and 50% coverage. The number of gene families in each partition is indicated</font></figcaption>
</figure>
</div>
</div>
</div>
<p>The number of clusters or gene families conserved (Shell+Persistent) increases as we go down in the taxonomical level. Note that the Shell partition shrinks at the species level, suggesting that the identity threshold is probably low at this level for a detailed pangenome analysis. However, as we will focus in the Persistent partition, the calculated pangenomes would be enough informative.</p>
<pre><code></code></pre>
</section>
<section id="integration-of-tnseq-and-pangenomes" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Integration of TnSeq and pangenomes</h1>
<p>As mentioned above, in order to compute the pangenomes, we carried out a new annotation of the genes in all the assemblies. This is required to have homogeneous gene definitions, but that also gives rise to some differences between the classical NCBI Genbank annotation that we used for the pangenome and the new annotations. Thus, in order to integrate the HB27 genes from TnSeq (TT_CXXXX) and pangenome genes (GCF_000008125.1_XXXXX) in the same table, we used the package <code>fuzzyjoin</code> that allows a flexible merge of the data, considering the gene start and stop as an interval rather than fixed coordinates. However, this flexibility also has a price, as we have some pangenome genes that matched with more than one gene in the classical annotation. We tested different levels of flexibility in the coordinates to maximize the number of genes in the table and at the same time minimize the number of duplicates.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#cross GCA_000008125.1 in all pangenomes</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>matrix <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>HB27 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pangenomes)){</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  matrix[[i]]<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">paste0</span>(<span class="st">"pangenomes/"</span>,pangenomes[i],<span class="st">"/table/GCF_000008125.1.tsv"</span>),<span class="at">header=</span><span class="cn">TRUE</span>,<span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>HB27 <span class="ot">&lt;-</span> <span class="fu">merge</span>(matrix[[<span class="dv">1</span>]][,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="dv">8</span>)],matrix[[<span class="dv">2</span>]][,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="dv">8</span>)],<span class="at">by=</span><span class="fu">c</span>(<span class="st">"gene"</span>,<span class="st">"contig"</span>,<span class="st">"start"</span>,<span class="st">"stop"</span>),<span class="at">all=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(HB27)[<span class="dv">5</span><span class="sc">:</span><span class="dv">6</span>] <span class="ot">&lt;-</span> pangenomes[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>HB27 <span class="ot">&lt;-</span> <span class="fu">merge</span>(HB27,matrix[[<span class="dv">3</span>]][,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="dv">8</span>)],<span class="at">by=</span><span class="fu">c</span>(<span class="st">"gene"</span>,<span class="st">"contig"</span>,<span class="st">"start"</span>,<span class="st">"stop"</span>),<span class="at">all=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>HB27 <span class="ot">&lt;-</span> <span class="fu">merge</span>(HB27,matrix[[<span class="dv">4</span>]][,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="dv">8</span>)],<span class="at">by=</span><span class="fu">c</span>(<span class="st">"gene"</span>,<span class="st">"contig"</span>,<span class="st">"start"</span>,<span class="st">"stop"</span>),<span class="at">all=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(HB27)[<span class="dv">7</span><span class="sc">:</span><span class="dv">8</span>] <span class="ot">&lt;-</span> pangenomes[<span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="co">#cross gene names </span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>tnseq <span class="ot">&lt;-</span> <span class="fu">read.csv2</span>(<span class="st">"tnSeq_full.csv"</span>)</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>HB27_genome <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"00_raw/refs/GCA_000008125.1.gtf"</span>,<span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,<span class="at">header=</span><span class="cn">FALSE</span>)</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>HB27_genome<span class="sc">$</span>Chr <span class="ot">&lt;-</span> <span class="st">"AE017221.1"</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>HB27_genome<span class="sc">$</span>Chr[<span class="dv">8088</span><span class="sc">:</span><span class="fu">nrow</span>(HB27_genome)] <span class="ot">&lt;-</span> <span class="st">"AE017222.1"</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>HB27_genome<span class="sc">$</span>V9 <span class="ot">&lt;-</span>  <span class="fu">substr</span>(HB27_genome<span class="sc">$</span>V9,<span class="dv">9</span>,<span class="dv">16</span>)</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(HB27_genome)[<span class="fu">c</span>(<span class="dv">4</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">9</span>)] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"start"</span>,<span class="st">"stop"</span>,<span class="st">"Genes"</span>)</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="co">#to reduce differnces in gene annotation, we use fuzzyjoin</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a><span class="co">#we use the highest max_dist value that don't give rise to duplicate gene names </span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> fuzzyjoin<span class="sc">::</span><span class="fu">difference_inner_join</span>(HB27[HB27<span class="sc">$</span>contig<span class="sc">==</span><span class="st">"contig_1"</span>,],HB27_genome[HB27_genome<span class="sc">$</span>V3<span class="sc">==</span><span class="st">"CDS"</span> <span class="sc">&amp;</span> HB27_genome<span class="sc">$</span>V1<span class="sc">==</span><span class="st">"AE017221.1"</span> ,<span class="fu">c</span>(<span class="dv">4</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">9</span>)], <span class="at">by =</span><span class="fu">c</span>(<span class="st">"start"</span>, <span class="st">"stop"</span>), <span class="at">max_dist =</span> <span class="dv">210</span> )</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>tmp2 <span class="ot">&lt;-</span> fuzzyjoin<span class="sc">::</span><span class="fu">difference_inner_join</span>(HB27[HB27<span class="sc">$</span>contig<span class="sc">==</span><span class="st">"contig_2"</span>,],HB27_genome[HB27_genome<span class="sc">$</span>V3<span class="sc">==</span><span class="st">"CDS"</span> <span class="sc">&amp;</span> HB27_genome<span class="sc">$</span>V1<span class="sc">==</span><span class="st">"AE017222.1"</span> ,<span class="fu">c</span>(<span class="dv">4</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">9</span>)], <span class="at">by =</span><span class="fu">c</span>(<span class="st">"start"</span>, <span class="st">"stop"</span>), <span class="at">max_dist =</span> <span class="dv">210</span> )</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>HB27_annotated <span class="ot">&lt;-</span> <span class="fu">rbind</span>(tmp,tmp2)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a><span class="co">#now we fetch the number of genomes with members in each family and paralogs</span></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a><span class="co">#load pangenome matrix</span></span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>matrix <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pangenomes)){</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>  matrix[[i]]<span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">paste0</span>(<span class="st">"pangenomes/"</span>,pangenomes[i],<span class="st">"/matrix.csv"</span>),<span class="at">sep=</span><span class="st">","</span>,<span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>  matrix[[i]] <span class="ot">&lt;-</span> matrix[[i]][,<span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">6</span>,<span class="fu">match</span>(<span class="st">"GCF_000008125.1"</span>,<span class="fu">names</span>(matrix[[i]])))]</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(matrix[[i]])[<span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="st">"gene"</span></span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a>  <span class="co">#HB27_annotated &lt;- merge(HB27_annotated,matrix[[i]], by="gene",all.x=TRUE)</span></span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(HB27_annotated)){</span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a>    k <span class="ot">&lt;-</span> <span class="fu">grep</span>(HB27_annotated<span class="sc">$</span>gene[j],matrix[[i]]<span class="sc">$</span>gene)</span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a>    HB27_annotated<span class="sc">$</span>isolates[j] <span class="ot">&lt;-</span> matrix[[i]]<span class="sc">$</span>No..isolates[k] </span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true" tabindex="-1"></a>    HB27_annotated<span class="sc">$</span>seqs[j] <span class="ot">&lt;-</span> matrix[[i]]<span class="sc">$</span>Avg.sequences.per.isolate[k] </span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(HB27_annotated)[<span class="fu">c</span>(<span class="fu">length</span>(HB27_annotated)<span class="sc">-</span><span class="dv">1</span>,<span class="fu">length</span>(HB27_annotated))] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">paste0</span>(pangenomes[i],<span class="st">"_isolates"</span>),<span class="fu">paste0</span>(pangenomes[i],<span class="st">"_seqs.per.isolate"</span>))</span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-44"><a href="#cb6-44" aria-hidden="true" tabindex="-1"></a>panTnseq <span class="ot">&lt;-</span> <span class="fu">merge</span>(tnseq,HB27_annotated[,<span class="fu">c</span>(<span class="dv">11</span>,<span class="dv">5</span><span class="sc">:</span><span class="dv">8</span>,<span class="dv">12</span><span class="sc">:</span><span class="dv">19</span>)])</span>
<span id="cb6-45"><a href="#cb6-45" aria-hidden="true" tabindex="-1"></a>panTnseq_Full <span class="ot">&lt;-</span>  <span class="fu">merge</span>(tnseq,HB27_annotated[,<span class="fu">c</span>(<span class="dv">11</span>,<span class="dv">5</span><span class="sc">:</span><span class="dv">8</span>,<span class="dv">12</span><span class="sc">:</span><span class="dv">19</span>)],<span class="at">all.x=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-46"><a href="#cb6-46" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(panTnseq_Full,<span class="st">"tnseq_pangenome2275.csv"</span>,<span class="at">sep=</span><span class="st">";"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We have been able to match in the pangenome 2190 genes out of 2275, with only 3 genes of the newly annotated HB27 (GCF_000008125.1_07995, GCF_000008125.1_08920, and GCF_000008125.1_08925) that matched with two genes in the classical annotation (TT_CXXXX).</p>
<p>The final table is saved as <a href="tnseq_pangenome2275.csv">tnseq_pangenome2275.csv</a> and <a href="tnseq_pangenome2275.xlsx.">tnseq_pangenome2275.xlsx.</a></p>
<section id="correlation-between-tnseq-and-pangenome" class="level2">
<h2 class="anchored" data-anchor-id="correlation-between-tnseq-and-pangenome">Correlation between TnSeq and Pangenome</h2>
<p>Now we are going to test the correlation between the TnSeq score and group classification of the HB27 genes, with the number of conserved genes in each pangenome and the pangenome cluster.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#corr</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>panTnseq_Full[,<span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">22</span><span class="sc">:</span><span class="dv">29</span>)] <span class="ot">&lt;-</span> <span class="fu">lapply</span>(panTnseq_Full[,<span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">22</span><span class="sc">:</span><span class="dv">29</span>)], as.numeric)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>M <span class="ot">&lt;-</span><span class="fu">cor</span>(panTnseq_Full[,<span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">22</span><span class="sc">:</span><span class="dv">29</span>)], <span class="at">method =</span> <span class="st">"spearman"</span>, <span class="at">use =</span> <span class="st">"pairwise.complete.obs"</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>testRes <span class="ot">&lt;-</span>  <span class="fu">cor.mtest</span>(panTnseq_Full[,<span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">22</span><span class="sc">:</span><span class="dv">29</span>)], <span class="at">conf.level =</span> <span class="fl">0.95</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">corrplot</span>(M, <span class="at">type=</span><span class="st">"lower"</span>, <span class="at">p.mat =</span> testRes<span class="sc">$</span>p, <span class="at">method =</span> <span class="st">'circle'</span>, <span class="at">insig=</span><span class="st">'blank'</span>,</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">tl.col=</span><span class="st">"black"</span>,<span class="at">tl.srt =</span> <span class="dv">45</span>, <span class="at">addCoef.col =</span><span class="st">'black'</span>, <span class="at">number.cex =</span> <span class="fl">0.5</span>, <span class="at">col=</span><span class="fu">brewer.pal</span>(<span class="at">n=</span><span class="dv">8</span>, <span class="at">name=</span><span class="st">"Spectral"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="panthermus_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption><font size="2">Figure 21. Correlation between TnSeq and Pangenome variables.</font></figcaption>
</figure>
</div>
</div>
</div>
<p>There’s no overall correlation.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co">#recuento</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>high <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>intermediate <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>less <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pangenomes)){</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  high[i] <span class="ot">&lt;-</span> <span class="fu">nrow</span>(<span class="fu">subset</span>(panTnseq,panTnseq[,<span class="dv">17</span><span class="sc">+</span>i]<span class="sc">==</span><span class="st">"persistent"</span>  <span class="sc">&amp;</span> panTnseq<span class="sc">$</span>cluster<span class="sc">==</span><span class="st">"Highly permissive"</span> ))</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>   intermediate[i] <span class="ot">&lt;-</span> <span class="fu">nrow</span>(<span class="fu">subset</span>(panTnseq,panTnseq[,<span class="dv">17</span><span class="sc">+</span>i]<span class="sc">==</span><span class="st">"persistent"</span> <span class="sc">&amp;</span> panTnseq<span class="sc">$</span>cluster<span class="sc">==</span><span class="st">"Intermediate"</span> ))</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>   less[i] <span class="ot">&lt;-</span> <span class="fu">nrow</span>(<span class="fu">subset</span>(panTnseq,panTnseq[,<span class="dv">17</span><span class="sc">+</span>i]<span class="sc">==</span><span class="st">"persistent"</span> <span class="sc">&amp;</span> panTnseq<span class="sc">$</span>cluster<span class="sc">==</span><span class="st">"Less permissive"</span> ))</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(high,intermediate,less)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(stats) <span class="ot">&lt;-</span> <span class="fu">names</span>(pangenomes)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">row.names</span>(stats),<span class="fu">stack</span>(stats))</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(stats) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"level"</span>,<span class="st">"ratio"</span>,<span class="st">"cluster"</span>)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(stats,<span class="fu">aes</span>(<span class="at">x=</span>level, <span class="at">y=</span>ratio,<span class="at">group=</span>cluster)) <span class="sc">+</span> </span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>         <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">fill=</span>cluster),<span class="at">stat =</span> <span class="st">"identity"</span>,<span class="at">position=</span><span class="st">"fill"</span>,<span class="at">color=</span><span class="st">"grey40"</span>, <span class="at">alpha=</span><span class="fl">0.8</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Persistent Genes  (%)"</span>)<span class="sc">+</span> <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>ratio),<span class="at">size=</span><span class="dv">5</span>, <span class="at">position =</span> <span class="fu">position_fill</span>(<span class="at">vjust=</span><span class="fl">0.5</span>) , <span class="at">col =</span> <span class="st">"black"</span>)<span class="sc">+</span><span class="fu">xlab</span>(<span class="st">"Pangenome level"</span>) <span class="sc">+</span><span class="fu">scale_y_continuous</span>(<span class="at">labels =</span> scales<span class="sc">::</span>percent) <span class="sc">+</span> <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="panthermus_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption><font size="2">Figure 22. Persistant genes in each pangenome).</font></figcaption>
</figure>
</div>
</div>
</div>
<p>Now, let’s see genes distribution in the different groups using Venn diagrams.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">Intermediate =</span> panTnseq_Full<span class="sc">$</span>Genes[panTnseq_Full<span class="sc">$</span>cluster<span class="sc">==</span><span class="st">"Intermediate"</span>], </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Highly_permissive =</span> panTnseq_Full<span class="sc">$</span>Genes[panTnseq_Full<span class="sc">$</span>cluster<span class="sc">==</span><span class="st">"Highly permissive"</span>], </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Less_permissive =</span> panTnseq_Full<span class="sc">$</span>Genes[panTnseq_Full<span class="sc">$</span>cluster<span class="sc">==</span><span class="st">"Less permissive"</span>], </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Persistent =</span> panTnseq_Full<span class="sc">$</span>Genes[panTnseq_Full<span class="sc">$</span>tt_i4c5<span class="sc">==</span><span class="st">"persistent"</span>]</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggvenn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: grid</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: scales</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'scales'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:purrr':

    discard</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:readr':

    col_factor</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>tt <span class="ot">&lt;-</span> <span class="fu">ggvenn</span>(</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  x, </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill_color =</span> <span class="fu">c</span>(<span class="st">"#EFC000FF"</span>,<span class="st">"#0073C2FF"</span>,<span class="st">"cornflowerblue"</span>, <span class="st">"#CD534CFF"</span>),</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">stroke_size =</span> <span class="fl">0.5</span>, <span class="at">set_name_size =</span> <span class="dv">4</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Thermus thermophilus"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))<span class="sc">+</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>   <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="fl">0.15</span>, <span class="fl">0.15</span>)))</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>x[[<span class="dv">4</span>]] <span class="ot">&lt;-</span> panTnseq_Full<span class="sc">$</span>Genes[panTnseq_Full<span class="sc">$</span>thermus_i4c5<span class="sc">==</span><span class="st">"persistent"</span>]</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>thermus <span class="ot">&lt;-</span> <span class="fu">ggvenn</span>(</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  x, </span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill_color =</span> <span class="fu">c</span>(<span class="st">"#EFC000FF"</span>,<span class="st">"#0073C2FF"</span>,<span class="st">"cornflowerblue"</span>, <span class="st">"#CD534CFF"</span>),</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">stroke_size =</span> <span class="fl">0.5</span>, <span class="at">set_name_size =</span> <span class="dv">4</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Thermus"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))<span class="sc">+</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>   <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="fl">0.15</span>, <span class="fl">0.15</span>)))</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>x[[<span class="dv">4</span>]] <span class="ot">&lt;-</span> panTnseq_Full<span class="sc">$</span>Genes[panTnseq_Full<span class="sc">$</span>thermaceae_i4c5<span class="sc">==</span><span class="st">"persistent"</span>]</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>thermaceae <span class="ot">&lt;-</span> <span class="fu">ggvenn</span>(</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>  x, </span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill_color =</span> <span class="fu">c</span>(<span class="st">"#EFC000FF"</span>,<span class="st">"#0073C2FF"</span>,<span class="st">"cornflowerblue"</span>, <span class="st">"#CD534CFF"</span>),</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">stroke_size =</span> <span class="fl">0.5</span>, <span class="at">set_name_size =</span> <span class="dv">4</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Thermaceae"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))<span class="sc">+</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>   <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="fl">0.15</span>, <span class="fl">0.15</span>)))</span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>x[[<span class="dv">4</span>]] <span class="ot">&lt;-</span> panTnseq_Full<span class="sc">$</span>Genes[panTnseq_Full<span class="sc">$</span>deinococcota_i4c5<span class="sc">==</span><span class="st">"persistent"</span>]</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>deinococcota <span class="ot">&lt;-</span> <span class="fu">ggvenn</span>(</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>  x, </span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill_color =</span> <span class="fu">c</span>(<span class="st">"#EFC000FF"</span>,<span class="st">"#0073C2FF"</span>,<span class="st">"cornflowerblue"</span>, <span class="st">"#CD534CFF"</span>),</span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">stroke_size =</span> <span class="fl">0.5</span>, <span class="at">set_name_size =</span> <span class="dv">4</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Deinococcota"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))<span class="sc">+</span></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>   <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="fl">0.15</span>, <span class="fl">0.15</span>)))</span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a><span class="fu">ggarrange</span>(tt, thermus, thermaceae, deinococcota, </span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a>          <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>,<span class="st">"C"</span>,<span class="st">"D"</span>), <span class="at">ncol=</span><span class="dv">2</span>,<span class="at">nrow =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="panthermus_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption><font size="2">Figure 23. Venn diagram of TnSeq genes categories and Persistant genes in each pangenome).</font></figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="enrichment-analysis" class="level2">
<h2 class="anchored" data-anchor-id="enrichment-analysis">Enrichment analysis</h2>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>kk <span class="ot">&lt;-</span> <span class="fu">enrichKEGG</span>(<span class="at">gene         =</span> <span class="fu">na.omit</span>(panTnseq_Full[panTnseq_Full<span class="sc">$</span>cluster<span class="sc">!=</span><span class="st">"Highly permissive"</span>  <span class="sc">&amp;</span> panTnseq_Full<span class="sc">$</span>thermaceae_i4c5<span class="sc">==</span><span class="st">"persistent"</span>,<span class="dv">1</span>]),</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">organism     =</span> <span class="st">'tth'</span>,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Reading KEGG annotation online: "https://rest.kegg.jp/link/tth/pathway"...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading KEGG annotation online: "https://rest.kegg.jp/list/pathway/tth"...</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dotplot</span>(kk, <span class="at">title=</span><span class="st">"Thermaceae core genome"</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>kk_D <span class="ot">&lt;-</span> <span class="fu">enrichKEGG</span>(<span class="at">gene         =</span> <span class="fu">na.omit</span>(panTnseq_Full[panTnseq_Full<span class="sc">$</span>cluster<span class="sc">!=</span><span class="st">"Highly permissive"</span>  <span class="sc">&amp;</span> panTnseq_Full<span class="sc">$</span>deinococcota_i4c5<span class="sc">==</span><span class="st">"persistent"</span>,<span class="dv">1</span>]),</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">organism     =</span> <span class="st">'tth'</span>,</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dotplot</span>(kk_D, <span class="at">title=</span><span class="st">"Deinococcota core genome"</span>)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggarrange</span>(p1, p2, </span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>), <span class="at">ncol=</span><span class="dv">1</span>,<span class="at">nrow =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="panthermus_files/figure-html/enrichment_pan-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption><font size="2">Figure 24. Significantly enriched KEGG Ontology groups made up of Intermediate/Less permissive genes within the Core genome (persistant genes) of Thermaceae (A) or Deinococcota (B).</font></figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>uniprot <span class="ot">&lt;-</span> <span class="fu">bitr_kegg</span>(panTnseq_Full<span class="sc">$</span>Genes, <span class="at">fromType=</span><span class="st">"kegg"</span>, <span class="at">toType=</span><span class="st">'uniprot'</span>, <span class="at">organism=</span><span class="st">'tth'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Reading KEGG annotation online: "https://rest.kegg.jp/conv/uniprot/tth"...</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>panTnseq_Full <span class="ot">&lt;-</span> panTnseq_Full <span class="sc">%&gt;%</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">uni =</span> <span class="fu">replace</span>(Genes, Genes <span class="sc">%in%</span> uniprot<span class="sc">$</span>kegg, uniprot<span class="sc">$</span>uniprot))</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>kk <span class="ot">&lt;-</span> <span class="fu">enrichKEGG</span>(<span class="at">gene         =</span> <span class="fu">na.omit</span>(panTnseq_Full[panTnseq_Full<span class="sc">$</span>cluster<span class="sc">!=</span><span class="st">"Highly permissive"</span>  <span class="sc">&amp;</span> panTnseq_Full<span class="sc">$</span>thermus_i4c5<span class="sc">==</span><span class="st">"persistent"</span>,<span class="dv">30</span>]),</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">organism     =</span> <span class="st">'tth'</span>, <span class="at">keyType =</span> <span class="st">'uniprot'</span>,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>           <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Reading KEGG annotation online: "https://rest.kegg.jp/conv/uniprot/tth"...</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dotplot</span>(kk, <span class="at">title=</span><span class="st">"Thermus core genome"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>kk_t <span class="ot">&lt;-</span> <span class="fu">enrichKEGG</span>(<span class="at">gene         =</span> <span class="fu">na.omit</span>(panTnseq_Full[panTnseq_Full<span class="sc">$</span>cluster<span class="sc">!=</span><span class="st">"Highly permissive"</span>  <span class="sc">&amp;</span> panTnseq_Full<span class="sc">$</span>thermaceae_i4c5<span class="sc">==</span><span class="st">"persistent"</span>,<span class="dv">30</span>]),</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">organism     =</span> <span class="st">'tth'</span>, <span class="at">keyType =</span> <span class="st">'uniprot'</span>,</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>           <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dotplot</span>(kk_t, <span class="at">title=</span><span class="st">"Thermaceae core genome"</span>)</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>kk_D <span class="ot">&lt;-</span> <span class="fu">enrichKEGG</span>(<span class="at">gene         =</span> <span class="fu">na.omit</span>(panTnseq_Full[panTnseq_Full<span class="sc">$</span>cluster<span class="sc">!=</span><span class="st">"Highly permissive"</span>  <span class="sc">&amp;</span> panTnseq_Full<span class="sc">$</span>deinococcota_i4c5<span class="sc">==</span><span class="st">"persistent"</span>,<span class="dv">30</span>]),</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>           <span class="at">organism     =</span> <span class="st">'tth'</span>, <span class="at">keyType =</span> <span class="st">'uniprot'</span>,</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>           <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>)</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">dotplot</span>(kk_D, <span class="at">title=</span><span class="st">"Deinococcota core genome"</span>)</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggarrange</span>(p1, p2, p3,</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>          <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>), <span class="at">ncol=</span><span class="dv">1</span>,<span class="at">nrow =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="panthermus_files/figure-html/enrichment_uniprot-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption><font size="2">Figure 25. Significantly enriched KEGG Ontology groups made up of Intermediate/Less permissive genes within the Core genome (persistant genes) of Thermaceae (A) or Deinococcota (B).</font></figcaption>
</figure>
</div>
</div>
</div>
<button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#collapseOne">
<p><em>Click for Session Information</em></p>
</button>
<div id="collapseOne" class="accordion-collapse collapse">

<div>


<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Europe/Madrid
tzcode source: internal

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ggvenn_0.1.10          scales_1.3.0           ggpubr_0.6.0          
 [4] clusterProfiler_4.12.0 RColorBrewer_1.1-3     corrplot_0.92         
 [7] ggstats_0.6.0          patchwork_1.2.0        lubridate_1.9.3       
[10] forcats_1.0.0          stringr_1.5.1          purrr_1.0.2           
[13] readr_2.1.5            tidyr_1.3.1            tibble_3.2.1          
[16] tidyverse_2.0.0        dplyr_1.1.4            DT_0.33               
[19] data.table_1.15.4      ggplot2_3.5.1          formatR_1.14          
[22] knitr_1.47            

loaded via a namespace (and not attached):
  [1] rstudioapi_0.16.0       jsonlite_1.8.8          magrittr_2.0.3         
  [4] farver_2.1.2            rmarkdown_2.27          fs_1.6.4               
  [7] zlibbioc_1.50.0         vctrs_0.6.5             memoise_2.0.1          
 [10] ggtree_3.12.0           rstatix_0.7.2           htmltools_0.5.8.1      
 [13] broom_1.0.6             gridGraphics_0.5-1      htmlwidgets_1.6.4      
 [16] plyr_1.8.9              cachem_1.1.0            igraph_2.0.3           
 [19] lifecycle_1.0.4         pkgconfig_2.0.3         fuzzyjoin_0.1.6        
 [22] Matrix_1.7-0            R6_2.5.1                fastmap_1.2.0          
 [25] gson_0.1.0              GenomeInfoDbData_1.2.12 digest_0.6.35          
 [28] aplot_0.2.3             enrichplot_1.24.0       colorspace_2.1-0       
 [31] AnnotationDbi_1.66.0    S4Vectors_0.42.0        RSQLite_2.3.7          
 [34] labeling_0.4.3          fansi_1.0.6             timechange_0.3.0       
 [37] abind_1.4-5             httr_1.4.7              polyclip_1.10-6        
 [40] compiler_4.4.0          bit64_4.0.5             withr_3.0.0            
 [43] backports_1.5.0         BiocParallel_1.38.0     carData_3.0-5          
 [46] viridis_0.6.5           DBI_1.2.3               ggforce_0.4.2          
 [49] ggsignif_0.6.4          MASS_7.3-61             HDO.db_0.99.1          
 [52] tools_4.4.0             ape_5.8                 scatterpie_0.2.3       
 [55] glue_1.7.0              nlme_3.1-165            GOSemSim_2.30.0        
 [58] shadowtext_0.1.3        reshape2_1.4.4          fgsea_1.30.0           
 [61] generics_0.1.3          gtable_0.3.5            tzdb_0.4.0             
 [64] hms_1.1.3               car_3.1-2               tidygraph_1.3.1        
 [67] utf8_1.2.4              XVector_0.44.0          BiocGenerics_0.50.0    
 [70] ggrepel_0.9.5           pillar_1.9.0            vroom_1.6.5            
 [73] yulab.utils_0.1.4       splines_4.4.0           tweenr_2.0.3           
 [76] treeio_1.28.0           lattice_0.22-6          bit_4.0.5              
 [79] tidyselect_1.2.1        GO.db_3.19.1            Biostrings_2.72.1      
 [82] gridExtra_2.3           IRanges_2.38.0          stats4_4.4.0           
 [85] xfun_0.45               graphlayouts_1.1.1      Biobase_2.64.0         
 [88] stringi_1.8.4           UCSC.utils_1.0.0        lazyeval_0.2.2         
 [91] ggfun_0.1.5             yaml_2.3.8              evaluate_0.24.0        
 [94] codetools_0.2-20        ggraph_2.2.1            qvalue_2.36.0          
 [97] ggplotify_0.1.2         cli_3.6.2               munsell_0.5.1          
[100] Rcpp_1.0.12             GenomeInfoDb_1.40.1     png_0.1-8              
[103] parallel_4.4.0          blob_1.2.4              DOSE_3.30.1            
[106] viridisLite_0.4.2       tidytree_0.4.6          crayon_1.5.2           
[109] rlang_1.1.4             cowplot_1.1.3           fastmatch_1.1-4        
[112] KEGGREST_1.44.0        </code></pre>
</div>
</div>


</div></div></section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>MRR, 2023</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://creativecommons.org/licenses/by-nc/2.0/">
      <i class="bi bi-badge-cc-fill" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/mredrejo">
      <i class="bi bi-twitter" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




<script src="site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>